public/scripts/project/lock.js

```diff
@@ -0,0 +1,65 @@
+// If the lock exists,it's almost guaranteed to be valid since the
+// expiry serves to prevent completely locking access to a project
+// in edge cases.
+define(function(require) {
+  var Constants = require('constants');
+
+  var LOCK_EXPIRY_MS = Constants.LOCK_EXPIRY_MS;
+  var LOCK_PREFIX = Constants.LOCK_PREFIX;
+
+  function getLockID(projectID) {
+    return LOCK_PREFIX + projectID;
+  }
+
+  function getLock(projectID) {
+    return localStorage.getItem(getLockID(projectID));
+  }
+
+  function setLock(projectID, value) {
+    localStorage.setItem(getLockID(projectID), value);
+  }
+
+  function removeLock(projectID) {
+    localStorage.removeItem(getLockID(projectID));
+  }
+
+  // Generates a lock on a project if one doesn't exist,
+  // or the one that exists has expired. Returns false
+  // if a valid lock already existed
+  function requestLock(projectID) {
+    function reacquire() {
+      setTimeout(reacquire, LOCK_EXPIRY_MS);
+      setLock(projectID, Date.now().toString());
+    }
+
+    var lock = readLock(projectID);
+    if (lock) {
+      return false;
+    }
+
+    reacquire();
+    return true;
+  }
+
+  // Returns the current valid lock, or undefined if there
+  // isn't one
+  function readLock(projectID) {
+    var lock = Number(getLock(projectID));
+    if (lock) {
+      // We add a buffer to ensure locked projects appear locked.
```

Comment:
Instead, you should store `Date.now()` before you call `getLock` and then compare. I don't think it's a great idea to pick an arbitrary time to add to the threshold